esmvalcore/preprocessor/_shared.py

```diff
@@ -517,3 +517,85 @@ def get_dims_along_coords(
     """Get a tuple with the dimensions along one or more coordinates."""
     dims = {d for coord in coords for d in _get_dims_along(cube, coord)}
     return tuple(sorted(dims))
+
+
+def apply_mask(
+    mask: np.ndarray | da.Array,
+    array: np.ndarray | da.Array,
+    dim_map: Iterable[int],
+) -> np.ma.MaskedArray | da.Array:
+    """Apply a (broadcasted) mask on an array.
+
+    Parameters
+    ----------
+    mask:
+        The mask to apply to array.
+    array:
+        The array to mask out.
+    dim_map : :class:`list`, :class:`tuple` etc
+        A mapping of the dimensions of *mask* to their corresponding
+        dimension in *array*. *dim_map* must be the same length as the
+        number of dimensions in *mask*. Each element of *dim_map*
+        corresponds to a dimension of *mask* and its value provides
+        the index in *array* which the dimension of *mask* corresponds
+        to, so the first element of *dim_map* gives the index of *array*
+        that corresponds to the first dimension of *mask* etc.
```

Comment:
this needs to be a bit more clear:
- is it a list/tuple of ints? or of lists/tuples?
- give an example e.g. `dim_map = [2, 3, 4]` -> `array[2]` will be masked with `mask[0]`?

Reply:
This is copy/pasted from the iris docs. I could also just link that? Anyway, this is not part of our public API.

Reply:
ha OK, pls post a link then - it's as clear as a foggy November day in the North Atlantic :grin: The module is private, but this seems to be a decent function for our public API? Up to you on that

Reply:
I changed it so the docstring now just refers to the `iris.util.broadcast_to_shape` function. That function has examples: https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.util.html#iris.util.broadcast_to_shape

Reply:
> The module is private, but this seems to be a decent function for our public API?

Let's keep it private for now, I'm not sure if this is the kind of functionality ESMValCore should provide.